src/NuGet.Protocol.Core.v3/ServiceIndexResourceV3Provider.cs

```diff
@@ -101,53 +110,74 @@ public ServiceIndexResourceV3Provider()
                 var utcNow = DateTime.UtcNow;
                 var entryValidCutoff = utcNow.Subtract(MaxCacheDuration);
 
-                ServiceIndexCacheInfo cacheInfo;
                 // check the cache before downloading the file
                 if (!_cache.TryGetValue(url, out cacheInfo) ||
                     entryValidCutoff > cacheInfo.CachedTime)
                 {
-                    var json = await GetIndexJson(source, token);
-
-                    // Use SemVer instead of NuGetVersion, the service index should always be
-                    // in strict SemVer format                    
-                    JToken versionToken;
-                    if (json.TryGetValue("version", out versionToken) &&
-                        versionToken.Type == JTokenType.String)
+                    // Track if the semaphore needs to be released
+                    var release = false;
+                    try
                     {
-                        SemanticVersion version;
-                        if (SemanticVersion.TryParse((string)versionToken, out version) &&
-                            version.Major == 3)
-                        {
-                            index = new ServiceIndexResourceV3(json, utcNow);
-                        }
-                        else
+                        // wait up to a minute for the lock, if we cannot get it just retrieve the file again.
+                        release = await _semaphore.WaitAsync(TimeSpan.FromMinutes(1), token);
```

Comment:
It could be removed. The logic is that requesting index.json should be a fairly cheap operation, if for some reason another thread is taking over a minute to complete the call, just get it a second time.